Added --ignore-alias to 'which' calls in setup.sh

cottsay opened the pull request:
On some systems, rm has a bash alias, often something like:
alias rm='rm -i'
/usr/bin/rm

This causes the setup.sh to not only fail to rm the file it intended, but also corrupt the alias that was present to begin with, causing rm to fail until the alias is fixed or the shell is restarted. Which has a --ignore-alias feature which will fix the issue.

cottsay committed:
Added --ignore-alias to 'which' calls in setup.sh

On some systems, rm has a bash alias, often something like:
alias rm='rm -i'
/usr/bin/rm

This causes the setup.sh to not only fail to rm the file it intended, but also corrupt the alias that was present to begin with, causing rm to fail until the alias is fixed or the sheel is restarted. Which has a --ignore-alias feature which will fix the issue.

dirk-thomas commented:
Could you please provide more information about the platform your are referring and what `which rm` returns for you?

cottsay commented:
I'm using Fedora 17. It looks like under root this alias is the default, but for some reason it is the default for my normal user, too:

```
cottsay@f17 ~ >>which rm
alias rm='rm -i'
    /usr/bin/rm
cottsay@f17 ~ >>
```

dirk-thomas commented:
The `--skip-alias` option is not supported on other platforms (e.g. on Ubuntu) so this patch can not be merged as-is.

Currently I don't see a cross-platform way to perform the which operation. Iterating over the PATH and searching the program might be the only fallback.

cottsay commented:
Wow, I never would have guessed that would happen, or I would have checked it myself. My apologies for not covering my bases before submitting a pull request.

Would a solution using grep be acceptable, something like:
`_RM=`which rm | grep -v alias``
...or is that not elegant enough?

dirk-thomas commented:
On OSX `which rm` does only output the alias line - not the binary path.
Can you give it shot with `-a` argument and check if that also works for you on Fedora?

`which -a rm | grep -v alias`

cottsay committed:
Change --ignore-alias to use grep/head instead

It looks like not all platforms support --skip-alias, so this solution should be more platform independent. grep is used to filter out any possible alias lines, while head is used to list only the first executable.

cottsay commented:
As requested:

```
cottsay@f17 ~ >>which -a rm | grep -v alias
    /usr/bin/rm
    /bin/rm
cottsay@f17 ~ >>
```

A possible solution to returning multiple paths using head:

```
cottsay@f17 ~ >>which -a rm | grep -v ^alias | head -1
    /usr/bin/rm
cottsay@f17 ~ >>
```

I also threw a caret before alias in case for SOME STRANGE REASON "alias" is in the path to rm. Thoughts on this?

Just FYI:

```
cottsay@f17 ~ >>which -v
GNU which v2.20, Copyright (C) 1999 - 2008 Carlo Wood.
GNU which comes with ABSOLUTELY NO WARRANTY;
This program is free software; your freedom to use, change
and distribute this program is protected by the GPL.
cottsay@f17 ~ >>
```

dirk-thomas commented:
This version also does not work on all platform. On some "which -a xxx" return a first line like:

```
xxx: aliased to xxx --args
```

dirk-thomas commented:
Since there seems to be no cross-platform way to use `which` I will modify the code to not use which at all. Instead invoke the commands directly with as specific arguments (even if default) as possible to override any modification made by custom aliases.

If a user still breaks one of the commands (so the aliased command does not do what it is supposed to do) that I will blame it to the custom alias and the user should revise that.
